general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
 #define MFEM_MEM_MANAGER_HPP
 
 #include "globals.hpp"
+#include "error.hpp"
+#include <cstring> // std::memcpy
+#include <type_traits> // std::is_const
 
 namespace mfem
 {
 
 // Implementation of MFEM's lightweight device/host memory manager designed to
 // work seamlessly with the OCCA, RAJA, and other kernels supported by MFEM.
 
+/// Memory types supported by MFEM.
+enum class MemoryType
+{
+   HOST,      ///< Host memory; using new[] and delete[]
+   HOST_32,   ///< Host memory alligned at 32 bytes
```

Comment:
These memory types can be implemented later -- we can add `TODO` in their description. One use case for this will be in the template code -- some data there needs to be allocated with proper alignment in order to fit with the vectorization intrinsics (this is one of the reasons that branch was never merged). Another reason is working with libCEED -- unless I'm mistaken, libCEED assumes that all vectors are allocated with 64-byte alignment which is `MemoryType::HOST_64`.

Reply:
I suspect it wouldn't cost much if all our Memory blocks were always 64-byte aligned. It could simplify the interface and would be one less thing to worry about.